superdesk/auth/oauth.py

```diff
@@ -24,53 +24,197 @@
 
 """
 
+import logging
+from datetime import datetime
+from typing import Optional, List
 import superdesk
+from flask import url_for, render_template
+from authlib.integrations.flask_client import OAuth
+from authlib.integrations.requests_client import OAuth2Session
+from authlib.oauth2.rfc6749.wrappers import OAuth2Token
+from bson import ObjectId
+from superdesk.resource import Resource
+from superdesk.services import BaseService
 
-from flask import url_for, session
-from flask_oauthlib.client import OAuth
+from superdesk.auth import auth_user, TEMPLATE
 
-from superdesk.auth import auth_user
 
+logger = logging.getLogger(__name__)
 
 bp = superdesk.Blueprint('oauth', __name__)
+oauth = None
+REFRESH_TOKEN_URL = "https://oauth2.googleapis.com/token"
 
 
 def init_app(app):
+    global oauth
     oauth = OAuth(app)
-    app.client_config['google_auth'] = False
-    if app.config.get('GOOGLE_CLIENT_ID') and app.config.get('GOOGLE_CLIENT_SECRET'):
-        app.client_config['google_auth'] = True
-        configure_google(app, oauth)
-        superdesk.blueprint(bp, app)
 
+    endpoint_name = 'oauth2_token'
+    service = OAuth2TokenService(endpoint_name, backend=superdesk.get_backend())
+    OAuth2TokenResource(endpoint_name, app=app, service=service)
 
-def configure_google(app, oauth):
-    google = oauth.remote_app(
+    app.client_config['google_auth'] = False
+    if app.config.get('GOOGLE_CLIENT_ID') and app.config.get('GOOGLE_CLIENT_SECRET'):
+        app.client_config['google_auth'] = bool(app.config.get('GOOGLE_LOGIN'))
+        if app.config.get('GOOGLE_GMAIL'):
+            extra_scopes = ['https://mail.google.com']
+            refresh = True
+        else:
+            extra_scopes = []
+            refresh = False
+
+        configure_google(app, extra_scopes=extra_scopes, refresh=refresh)
+
+
+class OAuth2TokenResource(Resource):
+    schema = {
+        'identifier': {
+            'type': 'string',
+            'unique': True,
+        },
+        'name': {
+            'type': 'string',
+            'required': True,
+            'nullable': False,
+            'empty': False,
+        },
+        'email': {
+            'type': 'string',
+        },
+        'access_token': {
+            'type': 'string',
+            'required': True,
+            'nullable': False,
+            'empty': False,
+        },
+        'refresh_token': {
+            'type': 'string',
+        },
+        'expires_at': {
+            'type': 'datetime'
+        },
+    }
+    internal_resource = True
+
+
+class OAuth2TokenService(BaseService):
+    """Store OAuth tokens"""
+
+
+def token2dict(
+    identifier: str,
+    email: str,
+    token: OAuth2Token,
+    name: str = "google",
+) -> dict:
```

Comment:
I think it's a time to add a type checker, afaik we don't use any right now.

Reply:
indeed, we should evaluate those available, but `mypy` is a pretty popular choice

Reply:
resolving it as you're integrating mypy in a separate ticket